src/src/Helpers/TarStream.cpp

```diff
@@ -0,0 +1,603 @@
+#include "../Helpers/TarStream.h"
+
+#if FEATURE_TARSTREAM_SUPPORT
+# include "../Globals/ESPEasy_time.h"
+# include "../Helpers/ESPEasy_Storage.h"
+# include "../Helpers/StringConverter.h"
+
+/**
+ * TarStream : Create/receive a .tar file while streaming via http webserver
+ * Copyright (c) 2023.. Ton Huisman for ESPEasy
+ *
+ * Changelog: See TarStream.h
+ */
+
+/**
+ * TarFileInfo_struct implementation
+ */
+TarFileInfo_struct::TarFileInfo_struct(const String _fileName,
+                                       size_t       _fileSize) :
+  fileName(_fileName), fileSize(_fileSize) {
+  tarSize = ((_fileSize % TAR_BLOCK_SIZE == 0 ? 0 : 1) + (_fileSize / TAR_BLOCK_SIZE)) * TAR_BLOCK_SIZE; // Multiple of block size
+  # if TAR_STREAM_DEBUG
+  addLog(LOG_LEVEL_INFO, strformat(F("Added file: %s, size: %d, tarSize: %d"), _fileName.c_str(), fileSize, tarSize));
+  # endif // if TAR_STREAM_DEBUG
+}
+
+/**
+ * TarStream implementation
+ */
+TarStream::TarStream() {}
+
+TarStream::TarStream(const String fileName)
+  : _fileName(fileName) {}
+
+TarStream::TarStream(const String fileName, FileDestination_e destination)
+  : _fileName(fileName), _destination(destination) {}
+
+TarStream::~TarStream() {
+  if (_filesList.size() > 0) {
+    _filesList.clear();
+  }
+}
+
+size_t TarStream::write(uint8_t ch) {
+  // TODO implement
+  addLogMove(LOG_LEVEL_ERROR, concat(F("TarStream: "), F("write(ch) NOT IMPLEMENTED YET.")));
+  return 1u;
+}
+
+size_t TarStream::write(const uint8_t *buf,
+                        size_t         size) {
+  size_t bufOffset  = 0u;   // Offset into the buffer
+  bool   stayInLoop = true; // To allow processing the rest of the data
+
+  while (stayInLoop) {
+    stayInLoop = false;
+
+    switch (_streamState) {
+      case TarStreamState_e::Initial: // Initial behaves like WritingHeader
+      case TarStreamState_e::WritingHeader:
+      {
+        if (_headerPosition == 0) {
+          clearHeader();
+        }
+        size_t toMove = std::min(size - bufOffset, TAR_HEADER_SIZE - _headerPosition);
+
+        memcpy(&_tarData[_headerPosition], &buf[bufOffset], toMove);
+        _headerPosition += toMove;
+        bufOffset       += toMove;
+
+        if (_headerPosition == TAR_HEADER_SIZE) {
+          bool isBlank = true;
+
+          for (size_t n = 0; n < TAR_HEADER_SIZE && isBlank; n++) {
+            isBlank &= (_tarData[n] == 0u);
+          }
+
+          if (isBlank) {
+            _headerPosition = 0;
+            _streamState    = TarStreamState_e::WritingFinal;
+            # if TAR_STREAM_DEBUG
+            addLog(LOG_LEVEL_INFO, concat(F("TarStream: Switch from "), F("Initial/WritingHeader to WritingFinal")));
```

Comment:
Check for active log level.

Reply:
That's kind of a theme here... 😆 will fix 'm.